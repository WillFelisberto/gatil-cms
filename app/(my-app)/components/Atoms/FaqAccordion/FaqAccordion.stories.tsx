import type { Meta, StoryObj } from '@storybook/react';

import { FAQAccordion } from './FaqAccordion';

const meta: Meta<typeof FAQAccordion> = {
  title: 'Atoms/FAQAccordion',
  component: FAQAccordion,
  tags: ['autodocs']
};

export default meta;
type Story = StoryObj<typeof FAQAccordion>;

export const Default: Story = {
  args: {
    faqs: [
      {
        question: 'O que é o apadrinhamento de gatos?',
        answer:
          'É uma forma de contribuir com os cuidados e o bem-estar de um gatinho resgatado, ajudando com os custos.'
      },
      {
        question: 'Posso apadrinhar mais de um gato?',
        answer:
          'Sim! Você pode adotar ou apadrinhar quantos gatos desejar, desde que consiga suprir suas necessidades com responsabilidade e carinho.'
      },
      {
        question: 'Como funciona o apadrinhamento?',
        answer:
          'Você escolhe um gatinho e contribui mensalmente com uma quantia para ajudar nos cuidados dele.'
      }
    ]
  }
};
