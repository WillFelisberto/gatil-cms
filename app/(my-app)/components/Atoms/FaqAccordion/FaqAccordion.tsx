'use client';

import { ChevronDown, Minus } from 'lucide-react';
import { useState } from 'react';

type FAQ = {
  question: string;
  answer: string;
};

type FAQAccordionProps = {
  faqs: FAQ[];
};

export const FAQAccordion = ({ faqs }: FAQAccordionProps) => {
  const [openIndex, setOpenIndex] = useState<number | null>(null);

  const toggle = (index: number) => {
    setOpenIndex(index === openIndex ? null : index);
  };

  return (
    <dl
      className="max-w-4xl mx-auto mt-10 border border-gray-300 rounded overflow-hidden divide-y divide-gray-200"
      data-testid="faq-accordion"
    >
      {faqs.map((faq, index) => {
        const isOpen = index === openIndex;
        const answerId = `faq-answer-${index}`;
        const questionId = `faq-question-${index}`;

        return (
          <div key={index}>
            <dt>
              <button
                onClick={() => toggle(index)}
                className="w-full flex justify-between items-center text-left px-6 py-4 hover:bg-gray-50 transition group cursor-pointer"
                aria-expanded={isOpen}
                aria-controls={answerId}
                id={questionId}
                data-testid={`faq-question-${index}`}
              >
                <span className="font-medium text-[#013274] group-hover:underline">
                  {faq.question}
                </span>
                {isOpen ? (
                  <Minus className="w-5 h-5 text-[#013274]" />
                ) : (
                  <ChevronDown className="w-5 h-5 text-[#013274]" />
                )}
              </button>
            </dt>
            <dd
              id={answerId}
              role="region"
              aria-labelledby={questionId}
              className={`transition-all duration-300 overflow-hidden px-6 ${
                isOpen ? 'max-h-40 pb-6' : 'max-h-0 pb-0'
              }`}
              data-testid={`faq-answer-${index}`}
            >
              <p className="text-gray-700 text-sm leading-relaxed">{faq.answer}</p>
            </dd>
          </div>
        );
      })}
    </dl>
  );
};
