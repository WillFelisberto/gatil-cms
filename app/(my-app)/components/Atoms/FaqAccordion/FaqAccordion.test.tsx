import { fireEvent, render, screen } from '@/tests/test-utils';

import { FAQAccordion } from './FaqAccordion';

const mockFaqs = [
  {
    pergunta: 'O que é apadrinhamento?',
    resposta: 'É uma forma de ajudar financeiramente um gatinho.'
  },
  {
    pergunta: 'Posso apadrinhar mais de um gato?',
    resposta: 'Sim! Você pode apadrinhar quantos gatos quiser.'
  }
];

describe('<FAQAccordion />', () => {
  it('should render all questions', () => {
    const { container } = render(<FAQAccordion faqs={mockFaqs} />);

    mockFaqs.forEach((faq, index) => {
      expect(screen.getByTestId(`faq-question-${index}`)).toHaveTextContent(faq.pergunta);
    });

    expect(container).toMatchSnapshot();
  });

  it('should expand answer when a question is clicked', () => {
    render(<FAQAccordion faqs={mockFaqs} />);

    const question = screen.getByTestId('faq-question-0');
    fireEvent.click(question);

    const answer = screen.getByTestId('faq-answer-0');
    expect(answer).toHaveTextContent(mockFaqs[0].resposta);
    expect(question).toHaveAttribute('aria-expanded', 'true');
  });

  it('should collapse answer when the same question is clicked again', () => {
    render(<FAQAccordion faqs={mockFaqs} />);

    const question = screen.getByTestId('faq-question-0');
    const answer = screen.getByTestId('faq-answer-0');

    fireEvent.click(question);
    expect(answer).toHaveTextContent(mockFaqs[0].resposta);
    expect(question).toHaveAttribute('aria-expanded', 'true');

    fireEvent.click(question);
    expect(answer.className).toMatch(/max-h-0/);
    expect(question).toHaveAttribute('aria-expanded', 'false');
  });

  it('should only show one answer at a time', () => {
    render(<FAQAccordion faqs={mockFaqs} />);

    const firstQuestion = screen.getByTestId('faq-question-0');
    const secondQuestion = screen.getByTestId('faq-question-1');
    const firstAnswer = screen.getByTestId('faq-answer-0');
    const secondAnswer = screen.getByTestId('faq-answer-1');

    fireEvent.click(firstQuestion);
    expect(firstAnswer).toHaveTextContent(mockFaqs[0].resposta);
    expect(firstAnswer.className).toMatch(/max-h-40/);

    fireEvent.click(secondQuestion);
    expect(secondAnswer).toHaveTextContent(mockFaqs[1].resposta);
    expect(secondAnswer.className).toMatch(/max-h-40/);
    expect(firstAnswer.className).toMatch(/max-h-0/);
  });
});
