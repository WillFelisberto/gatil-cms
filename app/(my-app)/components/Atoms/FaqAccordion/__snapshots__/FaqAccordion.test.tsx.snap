// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FAQAccordion /> should render all questions 1`] = `
<div>
  <dl
    class="max-w-4xl mx-auto mt-10 border border-gray-300 rounded overflow-hidden divide-y divide-gray-200"
    data-testid="faq-accordion"
  >
    <div>
      <dt>
        <button
          aria-controls="faq-answer-0"
          aria-expanded="false"
          class="w-full flex justify-between items-center text-left px-6 py-4 hover:bg-gray-50 transition group cursor-pointer"
          data-testid="faq-question-0"
          id="faq-question-0"
        >
          <span
            class="font-medium text-[#013274] group-hover:underline"
          >
            O que é apadrinhamento?
          </span>
          <svg
            class="lucide lucide-chevron-down w-5 h-5 text-[#013274]"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </dt>
      <dd
        aria-labelledby="faq-question-0"
        class="transition-all duration-300 overflow-hidden px-6 max-h-0 pb-0"
        data-testid="faq-answer-0"
        id="faq-answer-0"
        role="region"
      >
        <p
          class="text-gray-700 text-sm leading-relaxed"
        >
          É uma forma de ajudar financeiramente um gatinho.
        </p>
      </dd>
    </div>
    <div>
      <dt>
        <button
          aria-controls="faq-answer-1"
          aria-expanded="false"
          class="w-full flex justify-between items-center text-left px-6 py-4 hover:bg-gray-50 transition group cursor-pointer"
          data-testid="faq-question-1"
          id="faq-question-1"
        >
          <span
            class="font-medium text-[#013274] group-hover:underline"
          >
            Posso apadrinhar mais de um gato?
          </span>
          <svg
            class="lucide lucide-chevron-down w-5 h-5 text-[#013274]"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </dt>
      <dd
        aria-labelledby="faq-question-1"
        class="transition-all duration-300 overflow-hidden px-6 max-h-0 pb-0"
        data-testid="faq-answer-1"
        id="faq-answer-1"
        role="region"
      >
        <p
          class="text-gray-700 text-sm leading-relaxed"
        >
          Sim! Você pode apadrinhar quantos gatos quiser.
        </p>
      </dd>
    </div>
  </dl>
</div>
`;
